tests/database/test_header_db.py

```diff
@@ -94,18 +96,141 @@ def test_headerdb_persist_disconnected_headers(headerdb, genesis_header):
+    assert_headers_eq(headerdb.get_canonical_head(), pseudo_genesis)
+
+    headerdb.persist_header_chain(headers[:7])
+    assert headerdb.get_header_chain_gaps() == ((9, -1),)
+
+    for number in range(1, 9):
+        # Make sure we can lookup the headers by block number
+        header_by_number = headerdb.get_canonical_block_header_by_number(number)
+        assert header_by_number.block_number == headers[number - 1].block_number
+
+    # Make sure patching the hole does not affect what our current head is
+    assert_headers_eq(headerdb.get_canonical_head(), pseudo_genesis)
+
+
+def test_write_batch_that_patches_gap_and_adds_new_at_the_tip(headerdb, genesis_header):
+    headerdb.persist_header(genesis_header)
+    headers = mk_header_chain(genesis_header, length=10)
+
+    assert headerdb.get_header_chain_gaps() == ()
+    # Persist the checkpoint header with a trusted score
+    pseudo_genesis = headers[7]
+    headerdb.persist_checkpoint_header(pseudo_genesis, 154618822656)
+    assert headerdb.get_header_chain_gaps() == ((1, 7), (9, -1))
+    assert_headers_eq(headerdb.get_canonical_head(), pseudo_genesis)
+
+    headerdb.persist_header_chain(headers)
+    assert headerdb.get_header_chain_gaps() == ((11, -1),)
+
+    for number in range(1, len(headers)):
+        # Make sure we can lookup the headers by block number
+        header_by_number = headerdb.get_canonical_block_header_by_number(number)
+        assert header_by_number.block_number == headers[number - 1].block_number
+    # Make sure patching the hole does not affect what our current head is
+    assert_headers_eq(headerdb.get_canonical_head(), headers[-1])
+
+
+@pytest.mark.parametrize(
+    'written_headers, evolving_gaps',
+    (
+        (   # consecutive updates, then overwriting existing header
+            (1, 2, 1),
+            (
+                (GapChange.TailWrite, ((2, -1),)),
+                (GapChange.TailWrite, ((3, -1),)),
+                (GapChange.NoChange, ((3, -1),))
+            )
+        ),
+        (   # consecutive updates
+            (1, 2, 3),
+            (
+                (GapChange.TailWrite, ((2, -1),)),
+                (GapChange.TailWrite, ((3, -1),)),
+                (GapChange.TailWrite, ((4, -1),))
+            )
+        ),
+        (   # missing a single header in the middle
+            (1, 3),
+            ((GapChange.TailWrite, ((2, -1),)), (GapChange.NewGap, ((2, 2), (4, -1),)))
+        ),
+        (   # missing three headers in the middle
+            (1, 5),
+            ((GapChange.TailWrite, ((2, -1),)), (GapChange.NewGap, ((2, 4), (6, -1),)))
+        ),
+        (   # missing three headers in the middle, then patching center, dividing existing hole
+            (1, 5, 3),
+            (
+                (GapChange.TailWrite, ((2, -1),)),
+                (GapChange.NewGap, ((2, 4), (6, -1),)),
+                (GapChange.GapSplit, ((2, 2), (4, 4), (6, -1),))
```

Comment:
```py
                (GapChange.GapSplit, (((2, 2), (4, 4)), 6))
```
I just wanted to see what it looks like with the proposed data structure. I feel a little mixed about it, but still seems like a win.

Reply:
Yeah, I like that more. Especially because of the cleaner RLP handling but also because it feels we "abused" our previous datastructure a bit. There simply is not `BlockNumber` of `-1` possible.